functional_tests/keras/test_keras_subclassed_model.py

```diff
@@ -0,0 +1,40 @@
+import numpy as np
+import tensorflow as tf
+import wandb
+from wandb.keras import WandbCallback
+
+run = wandb.init(project="keras")
+
+
```

Comment:
How about using a fixed seed here and then explicitly checking the contents of the saved model/artifact in a few asserts at the end of this script? Just thinking, this might be an overkill.

Reply:
When I use the `--debug` flag, I get this dict:
```
INFO: exit= 0
{ ':wandb:artifacts': {'model-lovely-dawn-32': {'num': 2, 'type': 'model'}},
  ':wandb:portfolio_links': {},
  ':wandb:runs': [ { 'alerts': [],
                     'config': {},
                     'config_wandb': { 'cli_version': '0.12.16.dev1',
                                       'code_path': 'code/functional_tests/keras/test_keras_subclassed_model.py',
                                       'framework': 'keras',
                                       'is_jupyter_run': False,
                                       'is_kaggle_kernel': False,
                                       'python_version': '3.7.12',
                                       'start_time': 1650982985,
                                       't': { '1': [2, 3, 55],
                                              '2': [2, 3, 55],
                                              '3': [3, 8],
                                              '4': '3.7.12',
                                              '5': '0.12.16.dev1',
                                              '8': [5]}},
                     'exitcode': 0,
                     'files': { 'code/functional_tests/keras/test_keras_subclassed_model.py': { 'size': 942},
                                'conda-environment.yaml': {'size': 12070},
                                'config.yaml': {'size': 454},
                                'diff.patch': {'size': 1362},
                                'media/graph/graph_summary_c9172717e0c0fa694ed8.graph.json': { 'size': 381},
                                'output.log': {'size': 2466},
                                'requirements.txt': {'size': 4728},
                                'wandb-metadata.json': {'size': 753},
                                'wandb-summary.json': {'size': 470}},
                     'git': { 'commit': 'c8f3625ed8c0d468d41d51c217e5f4458c552f6f',
                              'remote': 'https://github.com/soumik12345/client'},
                     'group': '',
                     'history': [ { '_runtime': 18,
                                    '_step': 0,
                                    '_timestamp': 1650983003,
                                    'accuracy': 0.10000000149011612,
                                    'epoch': 0,
                                    'loss': 247.49560546875,
                                    'val_accuracy': 0.15000000596046448,
                                    'val_loss': 2.302488327026367},
                                  { '_runtime': 19,
                                    '_step': 1,
                                    '_timestamp': 1650983004,
                                    'accuracy': 0.10999999940395355,
                                    'epoch': 1,
                                    'loss': 2.302546501159668,
                                    'val_accuracy': 0.15000000596046448,
                                    'val_loss': 2.3022634983062744}],
                     'host': 'ayut-simple',
                     'job_type': '',
                     'metrics': {},
                     'name': '',
                     'notes': '',
                     'output': { 'stderr': [ '2022-04-26 14:23:12.146581: W '
                                             'tensorflow/stream_executor/platform/default/dso_loader.cc:64] '
                                             'Could not load dynamic library '
                                             "'libcuda.so.1'; dlerror: "
                                             'libcuda.so.1: cannot open shared '
                                             'object file: No such file or '
                                             'directory; LD_LIBRARY_PATH: '
                                             '/usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64',
                                             '2022-04-26 14:23:12.146646: W '
                                             'tensorflow/stream_executor/cuda/cuda_driver.cc:269] '
                                             'failed call to cuInit: UNKNOWN '
                                             'ERROR (303)',
                                             '2022-04-26 14:23:12.146672: I '
                                             'tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] '
                                             'kernel driver does not appear to '
                                             'be running on this host '
                                             '(ayut-simple): '
                                             '/proc/driver/nvidia/version does '
                                             'not exist',
                                             '2022-04-26 14:23:12.147201: I '
                                             'tensorflow/core/platform/cpu_feature_guard.cc:142] '
                                             'This TensorFlow binary is '
                                             'optimized with oneAPI Deep '
                                             'Neural Network Library (oneDNN) '
                                             'to use the following CPU '
                                             'instructions in '
                                             'performance-critical '
                                             'operations:  AVX2 FMA',
                                             'To enable them in other '
                                             'operations, rebuild TensorFlow '
                                             'with the appropriate compiler '
                                             'flags.',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             '\x1b[33mWARNING\x1b[39m The '
                                             'save_model argument by default '
                                             'saves the model in the HDF5 '
                                             'format that cannot save custom '
                                             'objects like subclassed models '
                                             'and custom layers. This behavior '
                                             'will be deprecated in a future '
                                             'release in favor of the '
                                             'SavedModel format. Meanwhile, '
                                             'the HDF5 model is saved as W&B '
                                             'files and the SavedModel as W&B '
                                             'Artifacts.',
                                             '2022-04-26 14:23:21.161710: I '
                                             'tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] '
                                             'None of the MLIR Optimization '
                                             'Passes are enabled (registered '
                                             '2)',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             '\x1b[32m\x1b[41mERROR\x1b[39m\x1b[49m '
                                             "Can't save model in the h5py "
                                             'format. The model will be saved '
                                             'as W&B Artifacts in the '
                                             'SavedModel format.',
                                             '2022-04-26 14:23:24.067492: W '
                                             'tensorflow/python/util/util.cc:348] '
                                             'Sets are not currently '
                                             'considered sequences, but this '
                                             'may change in the future, so '
                                             'consider avoiding using them.',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             'Adding directory to artifact '
                                             '(/home/ayusht/wandb-cli/client/functional_tests/keras/wandb/run-20220426_142305-2catnaoo/files/model-best)... '
                                             'Done. 0.1s',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             'Adding directory to artifact '
                                             '(/home/ayusht/wandb-cli/client/functional_tests/keras/wandb/run-20220426_142305-2catnaoo/files/model-best)... '
                                             'Done. 0.1s'],
                                 'stdout': [ 'Epoch 1/2',
                                             '4/4 '
                                             '[==============================] '
                                             '- 2s 71ms/step - loss: 247.4956 '
                                             '- accuracy: 0.1000 - val_loss: '
                                             '2.3025 - val_accuracy: 0.1500',
                                             'Epoch 2/2',
                                             '4/4 '
                                             '[==============================] '
                                             '- 0s 16ms/step - loss: 2.3025 - '
                                             'accuracy: 0.1100 - val_loss: '
                                             '2.3023 - val_accuracy: 0.1500']},
                     'program': './test_keras_subclassed_model.py',
                     'run_id': '2catnaoo',
                     'summary': { 'accuracy': 0.10999999940395355,
                                  'best_epoch': 1,
                                  'best_val_loss': 2.3022634983062744,
                                  'epoch': 1,
                                  'graph': { '_type': 'graph-file',
                                             'path': 'media/graph/graph_summary_c9172717e0c0fa694ed8.graph.json',
                                             'sha256': 'c9172717e0c0fa694ed800293d3019e42127def001df804e051ce98464ef8707',
                                             'size': 381},
                                  'loss': 2.302546501159668,
                                  'val_accuracy': 0.15000000596046448,
                                  'val_loss': 2.3022634983062744},
                     'summary_wandb': {'runtime': 20},
                     'tags': [],
                     'telemetry': { '1': [2, 3, 55],
                                    '2': [2, 3, 55],
                                    '3': [3, 8],
                                    '4': '3.7.12',
                                    '5': '0.12.16.dev1',
                                    '8': [5]}}],
  ':wandb:runs_len': 1,
  ':wandb:sentry_events': [],
  ':yea:exit': 0}
```

Not sure if I can check the content of the saved model/artifact. The only info that I have access to imo is: { ':wandb:artifacts': {'model-lovely-dawn-32': {'num': 2, 'type': 'model'}},`.

Reply:
I was thinking about explicit assert statements in the training script itself, as in read in the resulting model and check what's in there. But like i said, this might be an overkill -- up to you to decide whether to add that (if you see value in it).

Reply:
Thanks for the clarification here. I have modified the test to assert on the files logged as artifact. Can you please take a look? :)